test/server/guarddog_impl_test.cc

```diff
@@ -0,0 +1,156 @@
+  NiceMock<Configuration::MockMain> config_multikill_;
+  NiceMock<Stats::MockStore> fakestats_;
+  MockSystemTimeSource time_source_;
+  std::chrono::system_clock::time_point time_point_;
+};
+
+class GuardDogMissTest : public testing::Test {
+protected:
+  GuardDogMissTest() : config_miss(1, 1000, 0, 0), config_mega(1000, 1, 0, 0) {}
+
+  NiceMock<Configuration::MockMain> config_miss;
+  NiceMock<Configuration::MockMain> config_mega;
+  NiceMock<Stats::MockStore> stats_store;
+  MockSystemTimeSource time_source_;
+  std::chrono::system_clock::time_point time_point_;
+};
+
+TEST_F(GuardDogDeathTest, KillDeathTest) {
+  // Is it German for "The Function"? Almost...
+  auto die_function = [&]() -> void {
+    InSequence s;
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    GuardDogImpl gd(fakestats_, config_kill_, time_source_);
+    auto unpet_dog = gd.createWatchDog(0);
+    gd.forceCheckForTest();
+    time_point_ += std::chrono::milliseconds(500);
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    gd.forceCheckForTest();
+  };
+  // Why do it this way? Any threads must be started inside the death test
+  // statement and this is the easiest way to accomplish that.
+  EXPECT_DEATH(die_function(), "");
+}
+
+TEST_F(GuardDogDeathTest, MultiKillDeathTest) {
+  auto die_function = [&]() -> void {
+    InSequence s;
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    GuardDogImpl gd(fakestats_, config_multikill_, time_source_);
+    auto unpet_dog1 = gd.createWatchDog(0);
+    gd.forceCheckForTest();
+    auto unpet_dog2 = gd.createWatchDog(1);
+    gd.forceCheckForTest();
+    time_point_ += std::chrono::milliseconds(501);
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    gd.forceCheckForTest();
+  };
+  EXPECT_DEATH(die_function(), "");
+}
+
+TEST_F(GuardDogDeathTest, NearDeathTest) {
+  InSequence s;
+  EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+  GuardDogImpl gd(fakestats_, config_multikill_, time_source_);
+  auto unpet_dog = gd.createWatchDog(0);
+  auto pet_dog = gd.createWatchDog(1);
+  for (int i = 0; i < 6; i++) {
+    time_point_ += std::chrono::milliseconds(100);
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    pet_dog->touch();
+    gd.forceCheckForTest();
+  }
+}
+
+TEST_F(GuardDogMissTest, MissTest) {
+  EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+  GuardDogImpl gd(stats_store, config_miss, time_source_);
+  auto unpet_dog = gd.createWatchDog(0);
+  time_point_ += std::chrono::milliseconds(501);
+  EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+  gd.forceCheckForTest();
+  gd.stopWatching(unpet_dog);
+  unpet_dog = nullptr;
```

Comment:
OK, I'll just use real stats and check for the result rather than mock and check for activity.  I was trying to figure out how to do it with MockStore and it was indeed proving painful.